File: ts/web3js/src/instructions/eata/delegateEphemeralAta.ts
import {
  PublicKey,
  TransactionInstruction,
  SystemProgram,
  AccountMeta,
} from "@solana/web3.js";
import { EATA_PROGRAM_ID, DELEGATION_PROGRAM_ID } from "../../constants";
import {
  ephemeralAtaPdaWithBumpFromOwnerAndMint,
  delegateBufferPdaFromDelegatedAccountAndOwnerProgram,
  delegationRecordPdaFromDelegatedAccount,
  delegationMetadataPdaFromDelegatedAccount,
} from "../../pda";

export interface DelegateEphemeralAtaAccounts {
  payer: PublicKey;
  user: PublicKey;
  mint: PublicKey;
  validator?: PublicKey | null;
}

export function createDelegateEphemeralAtaInstruction(
  accounts: DelegateEphemeralAtaAccounts,
): TransactionInstruction {
  const [ephemeralAta, bump] = ephemeralAtaPdaWithBumpFromOwnerAndMint(
    accounts.user,
    accounts.mint,
  );
  const buffer = delegateBufferPdaFromDelegatedAccountAndOwnerProgram(
    ephemeralAta,
    EATA_PROGRAM_ID,
  );
  const delegationRecord =
    delegationRecordPdaFromDelegatedAccount(ephemeralAta);
  const delegationMetadata =
    delegationMetadataPdaFromDelegatedAccount(ephemeralAta);

  const keys: AccountMeta[] = [
    { pubkey: accounts.payer, isWritable: true, isSigner: true },
    { pubkey: ephemeralAta, isWritable: true, isSigner: false },
    { pubkey: EATA_PROGRAM_ID, isWritable: false, isSigner: false },
    { pubkey: buffer, isWritable: true, isSigner: false },
    { pubkey: delegationRecord, isWritable: true, isSigner: false },
    { pubkey: delegationMetadata, isWritable: true, isSigner: false },
    { pubkey: DELEGATION_PROGRAM_ID, isWritable: false, isSigner: false },
    { pubkey: SystemProgram.programId, isWritable: false, isSigner: false },
  ];

  const dataBuffer = Buffer.alloc(35);
  let offset = 0;
  dataBuffer[offset++] = 4; // discriminator
  dataBuffer[offset++] = bump;

  if (accounts.validator) {
    dataBuffer[offset++] = 1; // Some
    dataBuffer.set(accounts.validator.toBuffer(), offset);
    offset += 32;
  } else {
    dataBuffer[offset++] = 0; // None
  }

  return new TransactionInstruction({
    programId: EATA_PROGRAM_ID,
    keys,
    data: dataBuffer.subarray(0, offset),
  });
}

File: rust/sdk/src/spl/builders/delegate_ephemeral_ata.rs
use crate::pda::{
    delegate_buffer_pda_from_delegated_account_and_owner_program,
    delegation_metadata_pda_from_delegated_account, delegation_record_pda_from_delegated_account,
};

use crate::{
    compat,
    consts::ESPL_TOKEN_PROGRAM_ID,
    cpi::DELEGATION_PROGRAM_ID,
    spl::{EphemeralAta, EphemeralSplDiscriminator},
};

pub struct DelegateEphemeralAtaBuilder {
    pub payer: compat::Pubkey,
    pub user: compat::Pubkey,
    pub mint: compat::Pubkey,
    pub validator: Option<compat::Pubkey>,
}

impl DelegateEphemeralAtaBuilder {
    #[inline(always)]
    pub fn instruction(&self) -> compat::Instruction {
        let (eata, _eata_bump) = EphemeralAta::find_pda(&self.user, &self.mint);
        let delegation_buffer = delegate_buffer_pda_from_delegated_account_and_owner_program(
            &eata,
            &ESPL_TOKEN_PROGRAM_ID,
        );
        let delegation_record = delegation_record_pda_from_delegated_account(&eata);
        let delegation_metadata = delegation_metadata_pda_from_delegated_account(&eata);

        let mut data = Vec::with_capacity(33);
        data.push(EphemeralSplDiscriminator::DelegateEphemeralAta as u8);
        if let Some(validator) = self.validator {
            data.extend_from_slice(validator.as_ref());
        }

        compat::Instruction {
            program_id: ESPL_TOKEN_PROGRAM_ID,
            accounts: vec![
                compat::AccountMeta::new(self.payer, true),
                compat::AccountMeta::new(eata, false),
                compat::AccountMeta::new_readonly(ESPL_TOKEN_PROGRAM_ID, false),
                compat::AccountMeta::new(delegation_buffer, false),
                compat::AccountMeta::new(delegation_record, false),
                compat::AccountMeta::new(delegation_metadata, false),
                compat::AccountMeta::new_readonly(DELEGATION_PROGRAM_ID, false),
                compat::AccountMeta::new_readonly(
                    compat::Pubkey::new_from_array(solana_system_interface::program::ID.to_bytes()),
                    false,
                ),
            ],
            data,
        }
    }
}
